File: frontend/src/stores/family.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import type { Family, FamilyMember } from '@/types/models'
import * as familyApi from '@/api/family'

/**
 * 家庭状态管理 Store
 * 管理家庭列表、成员和相关操作
 */
export const useFamilyStore = defineStore('family', () => {
  // 状态
  const families = ref<Family[]>([])
  const currentFamily = ref<Family | null>(null)
  const members = ref<FamilyMember[]>([])
  const loading = ref(false)
  
  // 计算属性
  
  /**
   * 家庭数量
   */
  const familyCount = computed(() => families.value.length)
  
  /**
   * 当前家庭成员数量
   */
  const memberCount = computed(() => members.value.length)
  
  /**
   * 当前用户是否是家庭创建者
   * @param userId 当前用户 ID
   */
  const isOwner = computed(() => (userId: string) => {
    return currentFamily.value?.owner_id === userId
  })
  
  /**
   * 用户拥有的家庭列表
   * @param userId 当前用户 ID
   */
  const ownedFamilies = computed(() => (userId: string) => {
    return families.value.filter(family => family.owner_id === userId)
  })
  
  /**
   * 用户加入的家庭列表（非创建者）
   * @param userId 当前用户 ID
   */
  const joinedFamilies = computed(() => (userId: string) => {
    return families.value.filter(family => family.owner_id !== userId)
  })
  
  // 操作方法
  
  /**
   * 加载家庭列表
   * 从后端获取用户所属的所有家庭
   */
  async function loadFamilies() {
    loading.value = true
    try {
      const response = await familyApi.getFamilyList()
      if (response.success && response.data) {
        families.value = response.data
      }
      return response
    } finally {
      loading.value = false
    }
  }
  
  /**
   * 创建家庭
   * @param name 家庭名称
   */
  async function createFamily(name: string) {
    const response = await familyApi.createFamily(name)
    if (response.success) {
      // 重新加载家庭列表
      await loadFamilies()
    }
    return response
  }
  
  /**
   * 加入家庭
   * @param familyCode 家庭邀请码
   */
  async function joinFamily(familyCode: string) {
    const response = await familyApi.joinFamily(familyCode)
    if (response.success) {
      // 重新加载家庭列表
      await loadFamilies()
    }
    return response
  }
  
  /**
   * 离开家庭
   * @param familyId 家庭 ID
   */
  async function leaveFamily(familyId: string) {
    const response = await familyApi.leaveFamily(familyId)
    if (response.success) {
      // 从本地列表中移除
      families.value = families.value.filter(family => family._id !== familyId)
      // 如果离开的是当前家庭，清空当前家庭
      if (currentFamily.value?._id === familyId) {
        currentFamily.value = null
        members.value = []
      }
    }
    return response
  }
  
  /**
   * 加载家庭成员
   * @param familyId 家庭 ID
   */
  async function loadMembers(familyId: string) {
    loading.value = true
    try {
      const response = await familyApi.getFamilyMembers(familyId)
      if (response.success && response.data) {
        members.value = response.data
        // 更新当前家庭
        const family = families.value.find(f => f._id === familyId)
        if (family) {
          currentFamily.value = family
        }
      }
      return response
    } finally {
      loading.value = false
    }
  }
  
  /**
   * 移除家庭成员
   * @param familyId 家庭 ID
   * @param userId 用户 ID
   */
  async function removeMember(familyId: string, userId: string) {
    const response = await familyApi.removeFamilyMember(familyId, userId)
    if (response.success) {
      // 从本地成员列表中移除
      members.value = members.value.filter(member => member.user_id !== userId)
      // 更新家庭成员数量
      const family = families.value.find(f => f._id === familyId)
      if (family) {
        family.member_count = Math.max(0, family.member_count - 1)
      }
    }
    return response
  }
  
  /**
   * 更新家庭信息
   * @param familyId 家庭 ID
   * @param name 新名称
   */
  async function updateFamily(familyId: string, name: string) {
    const response = await familyApi.updateFamily(familyId, name)
    if (response.success) {
      // 更新本地家庭列表中的名称
      const family = families.value.find(f => f._id === familyId)
      if (family) {
        family.name = name
      }
      // 如果是当前家庭，也更新当前家庭
      if (currentFamily.value?._id === familyId) {
        currentFamily.value.name = name
      }
    }
    return response
  }
  
  /**
   * 删除家庭
   * @param familyId 家庭 ID
   */
  async function deleteFamily(familyId: string) {
    const response = await familyApi.deleteFamily(familyId)
    if (response.success) {
      // 从本地列表中移除
      families.value = families.value.filter(family => family._id !== familyId)
      // 如果删除的是当前家庭，清空当前家庭
      if (currentFamily.value?._id === familyId) {
        currentFamily.value = null
        members.value = []
      }
    }
    return response
  }
  
  /**
   * 设置当前家庭
   * @param familyId 家庭 ID
   */
  function setCurrentFamily(familyId: string) {
    const family = families.value.find(f => f._id === familyId)
    if (family) {
      currentFamily.value = family
    }
  }
  
  /**
   * 获取家庭共享冰箱
   * @param familyId 家庭 ID
   */
  async function getFamilyFridges(familyId: string) {
    return await familyApi.getFamilyFridges(familyId)
  }
  
  /**
   * 设置冰箱共享权限
   * @param fridgeId 冰箱 ID
   * @param isFamilyShared 是否家庭共享
   * @param isEditableByFamily 家庭成员是否可编辑
   */
  async function setFridgePermission(
    fridgeId: string,
    isFamilyShared: boolean,
    isEditableByFamily: boolean
  ) {
    return await familyApi.setFridgePermission(fridgeId, isFamilyShared, isEditableByFamily)
  }
  
  /**
   * 获取冰箱权限
   * @param fridgeId 冰箱 ID
   */
  async function getFridgePermission(fridgeId: string) {
    return await familyApi.getFridgePermission(fridgeId)
  }
  
  /**
   * 清空状态
   * 用户登出时调用
   */
  function clearState() {
    families.value = []
    currentFamily.value = null
    members.value = []
    loading.value = false
  }
  
  return {
    // 状态
    families,
    currentFamily,
    members,
    loading,
    // 计算属性
    familyCount,
    memberCount,
    isOwner,
    ownedFamilies,
    joinedFamilies,
    // 操作方法
    loadFamilies,
    createFamily,
    joinFamily,
    leaveFamily,
    loadMembers,
    removeMember,
    updateFamily,
    deleteFamily,
    setCurrentFamily,
    getFamilyFridges,
    setFridgePermission,
    getFridgePermission,
    clearState
  }
})

File: app/routes/family.py
# -*- coding: utf-8 -*-
"""家庭路由"""
from flask import Blueprint, request, jsonify
from app.services.family_service import FamilyService
from app.utils.jwt_auth import jwt_required
from app import db_client

family_bp = Blueprint('family', __name__, url_prefix='/family')


def get_db():
    """获取数据库实例"""
    return db_client.fridge


@family_bp.route('/create', methods=['POST'])
@jwt_required
def create_family():
    """创建家庭（JWT 认证）"""
    try:
        data = request.get_json()
        name = data.get('name', '').strip()
        
        if not name:
            return jsonify({'success': False, 'error': '家庭名称不能为空'}), 400
        
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        db = get_db()
        family_service = FamilyService(db)
        
        family = family_service.create_family(name, user_id)
        
        return jsonify({
            'success': True,
            'data': {
                'family_id': family._id,
                'family_code': family.family_code
            }
        }), 200
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/join', methods=['POST'])
@jwt_required
def join_family():
    """加入家庭（JWT 认证）"""
    try:
        data = request.get_json()
        family_code = data.get('family_code', '').strip().upper()
        
        if not family_code:
            return jsonify({'success': False, 'error': '家庭邀请码不能为空'}), 400
        
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        db = get_db()
        family_service = FamilyService(db)
        
        family_service.join_family(family_code, user_id)
        
        return jsonify({
            'success': True,
            'message': '成功加入家庭'
        }), 200
    except ValueError as e:
        return jsonify({'success': False, 'error': str(e)}), 400
    except Exception as e:
        return jsonify({'success': False, 'error': '服务器内部错误'}), 500


@family_bp.route('/leave/<family_id>', methods=['POST'])
@jwt_required
def leave_family(family_id):
    """离开家庭（JWT 认证）"""
    try:
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        db = get_db()
        family_service = FamilyService(db)
        
        family_service.leave_family(family_id, user_id)
        
        return jsonify({
            'success': True,
            'message': '已离开家庭'
        }), 200
    except ValueError as e:
        return jsonify({'success': False, 'error': str(e)}), 400
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/list', methods=['GET'])
@jwt_required
def list_families():
    """获取用户所在的所有家庭（JWT 认证）"""
    try:
        print(f"[调试] 收到家庭列表请求")
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        print(f"[调试] 用户ID: {user_id}")
        db = get_db()
        family_service = FamilyService(db)
        
        families = family_service.get_user_families(user_id)
        print(f"[调试] 查询到的家庭数量: {len(families)}")
        
        return jsonify({
            'success': True,
            'data': families
        })
    except Exception as e:
        print(f"[调试] 家庭列表查询出错: {str(e)}")
        import traceback
        traceback.print_exc()
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/<family_id>/members', methods=['GET'])
@jwt_required
def get_family_members(family_id):
    """获取家庭成员列表（JWT 认证）"""
    try:
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        db = get_db()
        family_service = FamilyService(db)
        
        # 检查是否是家庭成员
        if not family_service.is_family_member(family_id, user_id):
            return jsonify({'success': False, 'error': '您不是该家庭的成员'}), 403
        
        members = family_service.get_family_members(family_id)
        
        return jsonify({
            'success': True,
            'data': members
        }), 200
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/<family_id>', methods=['PUT'])
@jwt_required
def update_family(family_id):
    """更新家庭信息（JWT 认证）"""
    try:
        data = request.get_json()
        name = data.get('name', '').strip()
        
        if not name:
            return jsonify({'success': False, 'error': '家庭名称不能为空'}), 400
        
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        db = get_db()
        family_service = FamilyService(db)
        
        family_service.update_family(family_id, user_id, name)
        
        return jsonify({
            'success': True,
            'message': '家庭信息更新成功'
        }), 200
    except ValueError as e:
        return jsonify({'success': False, 'error': str(e)}), 403
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/<family_id>', methods=['DELETE'])
@jwt_required
def delete_family(family_id):
    """删除家庭（JWT 认证）"""
    try:
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        db = get_db()
        family_service = FamilyService(db)
        
        family_service.delete_family(family_id, user_id)
        
        return jsonify({
            'success': True,
            'message': '家庭已删除'
        }), 200
    except ValueError as e:
        return jsonify({'success': False, 'error': str(e)}), 403
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/<family_id>/members/<target_user_id>', methods=['DELETE'])
@jwt_required
def remove_member(family_id, target_user_id):
    """移除家庭成员（JWT 认证）"""
    try:
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        db = get_db()
        family_service = FamilyService(db)
        
        family_service.remove_member(family_id, user_id, target_user_id)
        
        return jsonify({
            'success': True,
            'message': '成员已移除'
        }), 200
    except ValueError as e:
        return jsonify({'success': False, 'error': str(e)}), 403
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/<family_id>/fridges', methods=['GET'])
@jwt_required
def get_family_fridges(family_id):
    """获取家庭共享的冰箱列表（JWT 认证）"""
    try:
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        db = get_db()
        family_service = FamilyService(db)
        
        # 检查是否是家庭成员
        if not family_service.is_family_member(family_id, user_id):
            return jsonify({'success': False, 'error': '您不是该家庭的成员'}), 403
        
        fridges = family_service.get_family_shared_fridges(family_id)
        
        return jsonify({
            'success': True,
            'data': fridges
        }), 200
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/fridge/<fridge_id>/permission', methods=['POST'])
@jwt_required
def set_fridge_permission(fridge_id):
    """设置冰箱权限（JWT 认证）"""
    try:
        data = request.get_json()
        is_family_shared = data.get('is_family_shared', False)
        is_editable_by_family = data.get('is_editable_by_family', False)
        
        print(f"[权限设置] 冰箱ID: {fridge_id}")
        print(f"[权限设置] 请求数据: {data}")
        print(f"[权限设置] 家庭共享: {is_family_shared}")
        print(f"[权限设置] 允许编辑: {is_editable_by_family}")
        
        # 从 JWT Token 中获取用户 ID
        user_id = request.user_id
        print(f"[权限设置] 用户ID: {user_id}")
        db = get_db()
        
        # 检查是否是冰箱所有者
        fridge = db.fridge.find_one({'_id': fridge_id})
        print(f"[权限设置] 冰箱信息: {fridge}")
        if not fridge or fridge['user_id'] != user_id:
            print(f"[权限设置] 权限检查失败 - 冰箱所有者: {fridge.get('user_id') if fridge else 'None'}, 当前用户: {user_id}")
            return jsonify({'success': False, 'error': '只有冰箱所有者可以设置权限'}), 403
        
        family_service = FamilyService(db)
        result = family_service.set_fridge_permission(fridge_id, is_family_shared, is_editable_by_family)
        print(f"[权限设置] 设置结果: {result}")
        
        return jsonify({
            'success': True,
            'message': '权限设置成功'
        }), 200
    except Exception as e:
        print(f"[权限设置] 异常: {str(e)}")
        import traceback
        traceback.print_exc()
        return jsonify({'success': False, 'error': str(e)}), 500


@family_bp.route('/fridge/<fridge_id>/permission', methods=['GET'])
@jwt_required
def get_fridge_permission(fridge_id):
    """获取冰箱权限（JWT 认证）"""
    try:
        print(f"[获取权限] 冰箱ID: {fridge_id}")
        db = get_db()
        family_service = FamilyService(db)
        permission = family_service.get_fridge_permission(fridge_id)
        print(f"[获取权限] 权限数据: {permission}")
        
        return jsonify({
            'success': True,
            'data': permission
        }), 200
    except Exception as e:
        print(f"[获取权限] 异常: {str(e)}")
        import traceback
        traceback.print_exc()
        return jsonify({'success': False, 'error': str(e)}), 500
